exporter/trace/jaeger/jaeger.go

```diff
@@ -0,0 +1,340 @@
+			Timestamp: time.Now().UnixNano() / 1000,
+			Fields:    fields,
+		})
+	}
+	//TODO: [rghetia] add links.
+	//
+	//var refs []*gen.SpanRef
+	//for _, link := range data.Links {
+	//	refs = append(refs, &gen.SpanRef{
+	//		TraceIdHigh: bytesToInt64(link.TraceID[0:8]),
+	//		TraceIdLow:  bytesToInt64(link.TraceID[8:16]),
+	//		SpanId:      bytesToInt64(link.SpanID[:]),
+	//	})
+	//}
+
+	return &gen.Span{
+		TraceIdHigh:   int64(data.SpanContext.TraceID.High),
+		TraceIdLow:    int64(data.SpanContext.TraceID.Low),
+		SpanId:        int64(data.SpanContext.SpanID),
+		ParentSpanId:  int64(data.ParentSpanID),
+		OperationName: data.Name, // TODO: if span kind is added then add prefix "Sent"/"Recv"
+		Flags:         int32(data.SpanContext.TraceOptions),
+		StartTime:     data.StartTime.UnixNano() / 1000,
+		Duration:      data.EndTime.Sub(data.StartTime).Nanoseconds() / 1000,
+		Tags:          tags,
+		Logs:          logs,
+		// TODO: goes with Links.
+		// References:    refs,
+	}
+}
+
+func attributeToTag(key string, a interface{}) *gen.Tag {
+	var tag *gen.Tag
+	switch value := a.(type) {
+	case bool:
+		tag = &gen.Tag{
+			Key:   key,
+			VBool: &value,
+			VType: gen.TagType_BOOL,
+		}
+	case string:
+		tag = &gen.Tag{
+			Key:   key,
+			VStr:  &value,
+			VType: gen.TagType_STRING,
+		}
+	case int64:
+		tag = &gen.Tag{
+			Key:   key,
+			VLong: &value,
+			VType: gen.TagType_LONG,
+		}
+	case int32:
+		v := int64(value)
+		tag = &gen.Tag{
+			Key:   key,
+			VLong: &v,
+			VType: gen.TagType_LONG,
+		}
+	case float64:
+		v := float64(value)
+		tag = &gen.Tag{
+			Key:     key,
+			VDouble: &v,
+			VType:   gen.TagType_DOUBLE,
+		}
+	}
+	return tag
+}
+
+// Flush waits for exported trace spans to be uploaded.
+//
+// This is useful if your program is ending and you do not want to lose recent spans.
+func (e *Exporter) Flush() {
+	e.bundler.Flush()
+}
+
+func (e *Exporter) upload(spans []*gen.Span) error {
+	batch := &gen.Batch{
+		Spans:   spans,
+		Process: e.process,
+	}
+	if e.endpoint != "" {
+		return e.uploadCollector(batch)
+	}
+	return e.uploadAgent(batch)
+}
+
+func (e *Exporter) uploadAgent(batch *gen.Batch) error {
+	return e.client.EmitBatch(batch)
+}
+
+func (e *Exporter) uploadCollector(batch *gen.Batch) error {
+	body, err := serialize(batch)
+	if err != nil {
+		return err
+	}
+	req, err := http.NewRequest("POST", e.endpoint, body)
+	if err != nil {
+		return err
+	}
+	if e.username != "" && e.password != "" {
+		req.SetBasicAuth(e.username, e.password)
+	}
+	req.Header.Set("Content-Type", "application/x-thrift")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	_, _ = io.Copy(ioutil.Discard, resp.Body)
+	resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return fmt.Errorf("failed to upload traces; HTTP status code: %d", resp.StatusCode)
+	}
+	return nil
+}
+
+func serialize(obj thrift.TStruct) (*bytes.Buffer, error) {
```

Comment:
I'm getting confused about the (?) different modes of output. Is there both a UDP sender and a HTTP sender? I was wondering when a thrift memory buffer would be re-used, vs. allocated for every send. Can you comment?

Reply:
There are two methods to export.
1. Via Agent - It uses UDP to send data to the agent.
2. Directly to Collector - It uses HTTP. (It could use GRPC).
It will only use one or the other. Second method is a preferred method.

I am not familiar with Thrift as much. I can look into it to check if there is a possibility of some optimization wrt memory buffer allocation. For now is it ok to put a TODO here?

Reply:
OK, that helps. I don't think a TODO is needed, I just wanted to be sure I knew what was happening.

Reply:
Probably too late but, I was wondering why not to use the `THTTPClient`(https://github.com/apache/thrift/blob/master/lib/go/thrift/http_client.go) instead of having an intermediate `TMemoryBuffer`.
For the reference: https://github.com/open-telemetry/opentelemetry-python/pull/508#discussion_r394515051.